[jayk.yang] 지하철 서비스 step3 리뷰 요청합니다.

tttuer opened the pull request:
부족한 부분 피드백 부탁드립니다. 감사합니다~!

tttuer committed:
feat: step1, step2 반영

step1 경로 조회 기능, step2 회원 인증 관련 기능(토큰 발급 API, 즐겨 찾기 기능), 테이블 스키마 추가

tttuer committed:
feat: step3 요금 계산 기능

거리별 요금 정책 및 노선별 추가 요금 계산 후 연령 별 할인 적용.

kakao-dino-bin committed:
refactor: 유저 인증, 검증 리팩토링

Null값으로 체크 하는 방어코드를 도메인에게 명령식으로 구현

-

kakao-dino-bin committed:
refactor: PathService 리팩토링

FareUtil, ShortestUtil로 최단경로, 경로의 요금 산정 리팩토링

-

kakao-dino-bin committed:
feat: 예외처리 수정 및 추가

RuntimeException으로 되어있던 구문을 CustomException으로 변경
ControllerAdvice를 사용해서 Exception을 핸들링

-

tttuer committed:
feat: CustomExceptionHandler 수정, Member email 패턴 수정

CustomExceptionHandler에서 올바르게 return을 보내주도록 수정, Member email 패턴이 맞지 않아 오류가 나던 부분 수정

tttuer committed:
feat: 상수들 빼내는 작업

findstar commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,27 @@
+        this.pathService = pathService;
+    }
+
     // TODO: 경로조회 기능 구현하기
```
구현이 완료되어 더 이상 사용하지 않는 주석은 정리해주세요 😄

findstar commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public PathResponse findShortestPath(LoginMember loginMember, Long source, Long target) {
+        List<Line> lines = lineDao.findAll();
```
List<Line>을 직접 다루지 않고 
일급 객체를 추가해서 사용해보면 어떨까요?

findstar commented on src/main/java/subway/util/FareUtil.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Comparator;
+import java.util.List;
+
+public class FareUtil {
```
요금을 구하는 로직을 FareUtil 로 정의하셨네요. 👍 
요금과 같은 경우에는 변동 가능성이 많기 때문에 
앞서 학습한 전략 패턴을 사용해 볼 수도 있습니다.

findstar commented on src/main/java/subway/member/domain/Member.java:
```diff
@@ -9,23 +16,26 @@
-        this.id = id;
+    public Member(String email, String password) {
+        String emailRule = EMAIL_RULE;
+        if (!email.matches(emailRule)) {
```
email 의 validate 유무를 생성자에서 확인하게 되면
모든 Member 객체의 생성 시점에 이러한 작업이 수행됩니다. 
email 과 같은 값은 새로운 Member 의 가입 / 정보 수정 시점에만 주요하게 관리할 수 있지 않을까요?

findstar commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,33 @@
+@Service
 public class AuthService {
+    private JwtTokenProvider jwtTokenProvider;
+    private MemberDao memberDao;
```
인증을 처리하는데에서는 Member 가 필요합니다.
MemberDao 를 직접 호출하지 말고 MemberService 를 사용해보면 어떨까요?

findstar commented on src/main/java/subway/member/application/MemberService.java:
```diff
@@ -28,6 +26,10 @@ public MemberResponse findMember(Long id) {
         return MemberResponse.of(member);
     }
 
+    public MemberResponse findMemberByEmail(String email) {
```
이 메서드는 호출하는 곳이 없습니다. 😄

findstar commented on src/main/java/subway/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,81 @@
+import subway.exception.*;
+
+@RestControllerAdvice
+public class CustomExceptionHandler {
```
CustomExceptionHandler 좋습니다. 💯

findstar commented on src/main/java/subway/line/domain/Sections.java:
```diff
@@ -40,15 +43,25 @@ public void addSection(Section section) {
+
+    public boolean isExist(Section section) {
+        for (Section section1 : sections) {
+            if (section1.getUpStation().equals(section.getUpStation()) && section1.getDownStation().equals(section.getDownStation())
+                    || section1.getUpStation().equals(section.getDownStation()) && section1.getDownStation().equals(section.getUpStation())) {
+                return true;
+            }
```
코드의 깊이가 깊고, 라인이 길어서 가독성이 떨어집니다. 별도의 메서드로 분리해보면 어떨까요?

findstar commented on src/main/java/subway/util/ShortestPathUtil.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class ShortestPathUtil {
```
Lines 를 기반으로하는 일급 객체를 활용하면 역할을 좀 더 도메인 영역으로 옮길 수 있습니다.

findstar reviewed (changes requested):
안녕하세요. 리뷰를 진행할 제임스입니다. 
미션의 구현을 잘 해주셨네요. 👍
몇가지 개선하면 좋을 사항에 대한 코멘트 남겼습니다. 
궁금한 부분은 언제든 질문 주세요! 😉
계속 화이팅입니다. 😄

tttuer committed:
refactor : 피드백 반영

주석 정리, List<Line> 을 Lines의 일급객체로 사용, FareUtil내의 할인 요금에 FareStrategy 적용, email validate를 Service에서 가입, 정보 수정 시점에 검사하도록 수정, 인증 처리시 MemberService를 사용하도록 수정
사용하지 않는 메서드 삭제, stream을 사용하여 깊이 짧도록 수정

tttuer committed:
refactor : isExist 내의 stream 줄바꿈 추가

findstar commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,43 @@
+        this.lineDao = lineDao;
+    }
+
+    public PathResponse findShortestPath(LoginMember loginMember, Long source, Long target) {
```
findShortestPath 메서드의 역할은 2가지인데요
1. 최단 경로를 구한다. 
2. 경로의 요금을 구한다 
이를 분리해보면 어떨까요?

findstar commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,43 @@
+    public PathResponse findShortestPath(LoginMember loginMember, Long source, Long target) {
+        Lines lines = new Lines(lineDao.findAll());
+
+        DijkstraShortestPath dijkstraShortestPath = lines.getDijkstraShortestPath();
```
DijkstraShortestPath 를 raw 하게 다루지 말고 Path 와 같은 도메인 객체를 만들어서 활용해보면 좋겠습니다

findstar commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -42,19 +43,19 @@ public ResponseEntity createMember(@RequestBody MemberRequest request) {
     }
 
     @GetMapping("/members/me")
```
컨트롤러에서 @RequestMapping("/members") 을 적용하면 동일한 구문을 줄일 수 있겠네요

findstar reviewed (changes requested):
안녕하세요 제임스 입니다.
리뷰가 늦어져 죄송합니다. 😭 
변경사항에 대한 코멘트 남겼습니다. 확인 부탁드려요. 

추가로 도메인 객체는 우리가 대상으로 하는 
문제 해결 영역에 대한 객체를 의미한다고 할 수 있습니다.
현재의 미션 과제는 지하철 서비스이기 때문에 
요금 계산 하는 로직에 도메인 객체를 좀 더 개선해 보시면 좋겠습니다.
궁금한점은 언제든지 DM 주세요.
그럼 계속 화이팅입니다. 😄

tttuer committed:
refactor: 피드백 반영

Member 경로 설정 변경, Path 도메인 생성

tttuer committed:
refactor: Path 수정

tttuer committed:
refactor: graph station, section 추가 분류

tttuer committed:
refactor: PathService에서 필요없는 것 삭제

findstar commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -25,17 +22,14 @@ public PathService(LineDao lineDao, StationDao stationDao) {
-    public PathResponse findShortestPath(LoginMember loginMember, Long source, Long target) {
-        Lines lines = new Lines(lineDao.findAll());
+    public PathResponse findMinDistance(LoginMember loginMember, Long source, Long target) {
+        Path path = new Path(new Lines(lineDao.findAll()));
```
DijkstraShortestPath 가 Path 도메인 객체로 처리되었네요.
Path 객체가 Lines 를 생성자의 인자로 전달 받습니다만. 조금더 풀어서 
```java
Graph graph = new Graph(lineDao.findAll());
Path path = graph.getPath(source, target); 
```
와 같은 형태가 되면 더 개선해볼 수 있을것 같네요.

findstar commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -11,50 +11,51 @@
 import java.net.URI;
 
 @RestController
+@RequestMapping("/members")
```
컨트롤러의 RequestMapping 으로 공통 영역이 처리되었네요 👍

findstar commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -25,17 +22,14 @@ public PathService(LineDao lineDao, StationDao stationDao) {
-                FareUtil.getMaxExtraFare(pathSections, lines.getLines()) + FareUtil.calculateDistanceFare(totalDistance));
-
-        return new PathResponse(shortestPath.stream()
+        return new PathResponse(path.getShortestPath(source,target).stream()
```
pathResponse 를 반환하기 위한 구조가 stream() 을 사용하더라도 복잡하다고 느껴지는데요, 
기회가 되면 더 개선할 수 있는 부분은 없을지 고민해봐주세요 😄

findstar reviewed (approved):
안녕하세요 제임스 입니다. 
피드백 반영 된 것 확인하였습니다. 
코멘트 남긴 부분은 필수는 아니지만 추가로 더 고민해봐주세요 😄  
이 PR 은 머지하겠습니다. 수고하셨습니다. 
다음 미션도 화이팅 입니다.👏
